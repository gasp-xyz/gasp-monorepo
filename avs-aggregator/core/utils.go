package core

import (
	"math/big"

	"github.com/Layr-Labs/eigensdk-go/crypto/bls"
	"github.com/ethereum/go-ethereum/accounts/abi"
	taskmanager "github.com/mangata-finance/eigen-layer-monorepo/avs-aggregator/bindings/FinalizerTaskManager"
	"golang.org/x/crypto/sha3"
)

// this hardcodes abi.encode() for taskmanager.IFinalizerTaskManagerTaskResponse
// unclear why abigen doesn't provide this out of the box...
func AbiEncodeTaskResponse(h *taskmanager.IFinalizerTaskManagerTaskResponse) ([]byte, error) {

	// The order here has to match the field ordering of taskmanager.IFinalizerTaskManagerTaskResponse
	taskResponseType, err := abi.NewType("tuple", "", []abi.ArgumentMarshaling{
		{
			Name: "referenceTaskIndex",
			Type: "uint32",
		},
		{
			Name: "BlockHash",
			Type: "bytes32",
		},
		{
			Name: "StorageProofHash",
			Type: "bytes32",
		},
	})
	if err != nil {
		return nil, err
	}
	arguments := abi.Arguments{
		{
			Type: taskResponseType,
		},
	}

	bytes, err := arguments.Pack(h)
	if err != nil {
		return nil, err
	}

	return bytes, nil
}

// GetTaskResponseDigest returns the hash of the TaskResponse, which is what operators sign over
func GetTaskResponseDigest(h *taskmanager.IFinalizerTaskManagerTaskResponse) ([32]byte, error) {

	encodeTaskResponseByte, err := AbiEncodeTaskResponse(h)
	if err != nil {
		return [32]byte{}, err
	}

	var taskResponseDigest [32]byte
	hasher := sha3.NewLegacyKeccak256()
	hasher.Write(encodeTaskResponseByte)
	copy(taskResponseDigest[:], hasher.Sum(nil)[:32])

	return taskResponseDigest, nil
}

// BINDING UTILS - conversion from contract structs to golang structs

// BN254.sol is a library, so bindings for G1 Points and G2 Points are only generated
// in every contract that imports that library. Thus the output here will need to be
// type casted if G1Point is needed to interface with another contract (eg: BLSPublicKeyCompendium.sol)
func ConvertToBN254G1Point(input *bls.G1Point) taskmanager.BN254G1Point {
	output := taskmanager.BN254G1Point{
		X: input.X.BigInt(big.NewInt(0)),
		Y: input.Y.BigInt(big.NewInt(0)),
	}
	return output
}

func ConvertToBN254G2Point(input *bls.G2Point) taskmanager.BN254G2Point {
	output := taskmanager.BN254G2Point{
		X: [2]*big.Int{input.X.A1.BigInt(big.NewInt(0)), input.X.A0.BigInt(big.NewInt(0))},
		Y: [2]*big.Int{input.Y.A1.BigInt(big.NewInt(0)), input.Y.A0.BigInt(big.NewInt(0))},
	}
	return output
}
