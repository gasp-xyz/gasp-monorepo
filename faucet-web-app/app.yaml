runtime: nodejs18
service: react-faucet

handlers:
  - url: /static/js/(.*)
    static_files: build/static/js/\1
    upload: build/static/js/(.*)
    secure: always
  - url: /static/css/(.*)
    static_files: build/static/css/\1
    upload: build/static/css/(.*)
    secure: always
  - url: /static/media/(.*)
    static_files: build/static/media/\1
    upload: build/static/media/(.*)
    secure: always
  - url: /(.*\.(json|ico|png|jpg|svg))$
    static_files: build/\1
    upload: build/.*\.(json|ico|png|jpg|svg)$
    secure: always
  - url: /
    static_files: build/index.html
    upload: build/index.html
    secure: always
  - url: /.*
    static_files: build/index.html
    upload: build/index.html
    secure: always