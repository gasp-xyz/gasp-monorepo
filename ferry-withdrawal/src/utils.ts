export function isEqual(first: Uint8Array, second: Uint8Array): boolean {
	if (first.length !== second.length) {
		return false;
	}
	return first.every((value, index) => value === second[index]);
}

export function maxBigInt(...args: bigint[]) {
	return BigInt(args.reduce((min, current) => (current > min ? current : min)));
}

export function minBigInt(...args: bigint[]) {
	return BigInt(args.reduce((min, current) => (current < min ? current : min)));
}

export async function asyncFilter<T>(arr: T[], predicate: any) {
	const results = await Promise.all(arr.map(predicate));
	return arr.filter((v: any, index: any) => {
		return results[index];
	});
}
