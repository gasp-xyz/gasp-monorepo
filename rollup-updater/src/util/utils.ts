import * as util from "node:util";
import type {
    FrameSystemEventRecord,
    PalletRolldownMessagesL1Update,
} from "@polkadot/types/lookup"
import type { Option } from "@polkadot/types-codec";


export function getMinRequestId(l2Update: any) {
    let minId = Math.min.apply(null,
        [
            l2Update[0].withdrawals,
            l2Update[0].cancels,
            l2Update[0].results
        ].flat()
            .map(function(item: any) {
                return Number(item.requestId.id);
            }))

    if (minId === Infinity) {
        return null;
    } else {
        return minId;
    }
}

function print(data: any) {
    console.log(util.inspect(data, { depth: null }));
}

export {
    print
}