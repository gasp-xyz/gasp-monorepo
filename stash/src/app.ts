import 'gasp-types'
import express, { Request, Response, json, urlencoded } from 'express'
import swaggerUI from 'swagger-ui-express'
import cors from 'cors'

import * as priceDiscoveryController from './controller/PriceDiscoveryController.js'
import * as tokenController from './controller/TokenController.js'
import * as xcmController from './controller/XcmController.js'
import * as coingeckoController from './controller/coingeckoListingController.js'
import * as coinmarketcapController from './controller/coinmarketcapListingController.js'
import * as tokenListController from './controller/tokenListController.js'
import * as stakingAprController from './controller/stakingAprController.js'
import * as liquidStakingController from './controller/liquidStakingController.js'
import * as networkController from './controller/networkController.js'
import * as tokenNetworkPortfolioController from './controller/tokenNetworkPortfolioController.js'
import * as faucetController from './controller/FaucetController.js'
import * as tracingController from './controller/TracingController.js'

import { createRequire } from 'module'
const require = createRequire(import.meta.url)
const swaggerFile = require('../swagger-output.json')

const app = express()

// App configuration
app.set('port', process.env.PORT || 8080)
app.use(json())
app.use(urlencoded({ extended: true }))
app.use(cors())

//Root (Health Check)
app.get('/', async (req: Request, res: Response): Promise<void> => {
  res.json({})
})

// Tokens
app.get('/token/order-buckets', tokenController.listTokenOrderBuckets)
// Token list
app.get('/token/list/stats', tokenListController.tokenList)
app.get('/token/:id/stats', tokenListController.tokenDetails)

// XCM
app.get('/xcm/channels', xcmController.channels)
app.get('/xcm/network', xcmController.network)
app.get('/xcm/tokens', xcmController.tokens)

app.get(
  '/collator/:collatorAddress/staking/dailyReward',
  stakingAprController.dailyReward
)
app.get('/collator/:collatorAddress/staking/apy', stakingAprController.apy)
app.get('/collators/staking/apy', stakingAprController.collatorsApy)

app.get(
  '/account/:address/liquid-staking/rewards-history/24h/sum',
  liquidStakingController.rewards24hours
)
app.get(
  '/account/:address/liquid-staking/rewards-history/month/sum',
  liquidStakingController.rewardsMonth
)

// Price Discovery API
app.get('/price-discovery/:currencyId/', priceDiscoveryController.getPrice)
app.get(
  '/price-history/pair/:baseCurrencyId/:targetCurrencyId',
  priceDiscoveryController.getHistoryPair
)
app.get('/price-history/:currencyId/', priceDiscoveryController.getHistory)
app.get(
  '/volume-history/pools/:currencyId/',
  priceDiscoveryController.getTradesPool
)
app.get('/volume-history/:currencyId/', priceDiscoveryController.getTradesAsset)
app.get(
  '/tvl-history/pools/:currencyId/',
  priceDiscoveryController.getVolumePool
)
app.get('/tvl-history/:currencyId/', priceDiscoveryController.getVolumeAsset)

// CoinGecko listing endpoints
app.get('/coingecko/pairs', coingeckoController.pairs)
app.get('/coingecko/tickers', coingeckoController.tickers)

app.get('/affirmed-network/list', networkController.networkList)
app.get('/affirmed-token/list', networkController.tokenList)
app.get(
  '/account/:address/token-portfolio',
  tokenNetworkPortfolioController.tokenNetworkPortfolio
)

//Faucet endpoint
app.get(
  '/faucet/requestTokens/:toAddress/captcha/:captcha/',
  faucetController.captcha
)

// Tracing endpoints
app.post('/tracing/tx/start', tracingController.startTracing)

app.get(
  '/tracing/tx/:txHashOrEntityId',
  tracingController.getTransactionStatusByTxHashOrEntityId
)

app.get(
  '/tracing/tx/listByAddress/:address',
  tracingController.getAllTransactionsByAddress
)

app.get(
  '/tracing/tx/listByAddress/:address/:status',
  tracingController.getAllTransactionsByAddressAndStatus
)

app.get(
  '/tracing/tx/findByEntityId/:entityId',

  tracingController.getATransactionByEntityId
)

// Coinmarketcap listing endpoints
app.get('/coinmarketcap/v1/summary', coinmarketcapController.summary)

app.use('/doc', swaggerUI.serve, swaggerUI.setup(swaggerFile))

export default app
