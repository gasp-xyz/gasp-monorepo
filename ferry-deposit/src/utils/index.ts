import util from "node:util";
import { type ApiPromise, Keyring } from "@polkadot/api";
import type { ApiDecoration } from "@polkadot/api/types";
import type { KeyringPair } from "@polkadot/keyring/types";
import type { StorageKey } from "@polkadot/types";
import type { Option, u32, u128 } from "@polkadot/types-codec";
import type {
	FrameSystemEventRecord,
	MangataTypesAssetsL1Asset,
	OrmlTokensAccountData,
	PalletRolldownMessagesChain,
	PalletRolldownMessagesDeposit,
	PalletRolldownMessagesL1Update,
	SpRuntimeAccountAccountId20,
} from "@polkadot/types/lookup";
import type { ITuple } from "@polkadot/types/types";
import { Mangata, type MangataGenericEvent, signTx } from "gasp-sdk";
import {
	http,
	type PrivateKeyAccount,
	createWalletClient,
	webSocket,
} from "viem";
import { type PublicClientConfig, createPublicClient } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { anvil } from "viem/chains";

import "gasp-types";

import type { H256 } from "@polkadot/types/interfaces/runtime";
import { hexToU8a, u8aToHex } from "@polkadot/util";
import type { KeypairType } from "@polkadot/util-crypto/types";
import { type PublicClient, encodeAbiParameters, keccak256 } from "viem";
import {
	ABI,
	L1_CHAIN,
	MANGATA_CONTRACT_ADDRESS,
	MIN_PROFIT,
} from "../common/constants.js";

function getL1ChainType(api: ApiPromise): PalletRolldownMessagesChain {
	return api.createType("PalletRolldownMessagesChain", L1_CHAIN);
}

function createL1Asset(
	api: ApiPromise,
	tokenAddress: Uint8Array,
): MangataTypesAssetsL1Asset {
	const chain: PalletRolldownMessagesChain = getL1ChainType(api);
	if (chain.isEthereum) {
		return api.createType("MangataTypesAssetsL1Asset", {
			Ethereum: tokenAddress,
		});
	} else if (chain.isArbitrum) {
		return api.createType("MangataTypesAssetsL1Asset", {
			Arbitrum: tokenAddress,
		});
	} else {
		throw new Error(`Unknown chain id ${chain.toHuman()}`);
	}
}

async function asyncFilter(arr: Deposit[], predicate: any) {
	const results = await Promise.all(arr.map(predicate));
	return arr.filter((v: any, index: any) => {
		return results[index];
	});
}

interface Deposit {
	readonly requestId: bigint;
	readonly depositRecipient: Uint8Array;
	readonly tokenAddress: Uint8Array;
	readonly amount: bigint;
	readonly timeStamp: bigint;
	readonly ferryTip: bigint;
}

class Ferry {
	l1: L1Interface;
	l2: L2Interface;
	me: Uint8Array;
	txCost: bigint;
	minProfit: bigint;

	constructor(
		me: Uint8Array,
		l1: L1Interface,
		l2: L2Interface,
		txCost: bigint,
		minProfit: bigint,
	) {
		this.me = me;
		this.l1 = l1;
		this.l2 = l2;
		this.txCost = txCost;
		this.minProfit = minProfit;
	}

	logFilteredOut(before: Deposit[], after: Deposit[], message: string) {
		const diff = before.length - after.length;
		if (diff > 0) {
			console.info(`filtered out ${diff} : ${message}`);
		}
	}

	async getPendingDeposits(): Promise<Deposit[]> {
		const end = await this.l1.getLatestRequestId();
		let start = await this.l2.getLastProcessedRequestId();
		if (end === null) {
			return Promise.resolve([]);
		}
		if (start === 0n) {
			start = 1n;
		}
		const deposits = await this.l1.getDeposits(start, end);

		const ferryableDeposits = await asyncFilter(
			deposits,
			async (elem: Deposit) => {
				const hash = await this.l1.getDepostiHash(elem.requestId);
				const isFerried = await this.l2.isFerried(hash);
				const isExecuted = await this.l2.isExecuted(elem.requestId);
				return !isFerried && !isExecuted;
			},
		);
		return ferryableDeposits;
	}

	async rateDeposits(deposits: Deposit[]): Promise<Deposit[]> {
		const nativeTokenAddress = await this.l2.getNativeTokenAddress();
		const balances = await this.l2.getBalances(this.me);

		const tokenAddressToBalance = new Map<string, bigint>(
			Array.from(balances, ([k, v]) => [u8aToHex(k), v]),
		);

		const validDeposits = deposits.filter((deposit) => {
			return deposit.amount > deposit.ferryTip;
		});
		this.logFilteredOut(deposits, validDeposits, "invalid deposit");

		const affordableDeposits = validDeposits.filter((deposit) => {
			const transferAmount = deposit.amount - deposit.ferryTip;
			const tokenAddress = u8aToHex(deposit.tokenAddress);

			if (tokenAddressToBalance.has(tokenAddress)) {
				if (deposit.tokenAddress === nativeTokenAddress) {
					return (
						tokenAddressToBalance.get(tokenAddress)! >=
						transferAmount + this.txCost
					);
				} else {
					return tokenAddressToBalance.get(tokenAddress)! >= transferAmount;
				}
			} else {
				return false;
			}
		});

		this.logFilteredOut(
			validDeposits,
			affordableDeposits,
			`not enought tokens to cover Account: ${u8aToHex(this.me)}`,
		);

		const ratedDeposits = affordableDeposits.sort((a, b) => {
			if (a.ferryTip > b.ferryTip) {
				return -1;
			} else if (a.ferryTip === b.ferryTip) {
				if (a.amount > b.amount) {
					return 1;
				} else if (a.amount > b.amount) {
					return 0;
				} else {
					return -1;
				}
			} else {
				return 1;
			}
		});

		const aboveProfitThreshold = await asyncFilter(
			ratedDeposits,
			async (elem: Deposit) => {
				return (
					(await this.l2.valutateToken(elem.tokenAddress, elem.ferryTip)) >=
					this.minProfit
				);
			},
		);

		this.logFilteredOut(
			affordableDeposits,
			aboveProfitThreshold,
			"profit below expected threshold",
		);

		return aboveProfitThreshold;
	}
}

async function dummyDeposit(uri: string) {
	const ANVIL_TEST_ACCOUNT =
		"0x8b3a350cf5c34c9194ca85829a2df0ec3153be0318b5e2d3348e872092edffba";
	const transport = webSocket(uri, { retryCount: 5 });
	const publicClient = createPublicClient({
		transport,
	});

	const acc: PrivateKeyAccount = privateKeyToAccount(ANVIL_TEST_ACCOUNT);

	const { request } = await publicClient.simulateContract({
		account: acc,
		address: MANGATA_CONTRACT_ADDRESS,
		abi: ABI,
		functionName: "deposit_native",
		value: BigInt(123456789),
	});

	const wc = createWalletClient({
		account: acc,
		chain: anvil,
		transport,
	});
	return await wc.writeContract(request);
}

interface L1Interface {
	isRolldownDeployed(): Promise<boolean>;
	getLatestRequestId(): Promise<bigint | null>;
	getDeposits(rangeStart: bigint, rangeEnd: bigint): Promise<Deposit[]>;
	getDepostiHash(requestId: bigint): Promise<Uint8Array>;
}

interface L2Interface {
	getBalances(address: Uint8Array): Promise<Map<Uint8Array, bigint>>;
	getLastProcessedRequestId(): Promise<bigint>;
	getLastProcessedRequestId(): Promise<bigint>;
	isExecuted(depositId: bigint): Promise<boolean>;
	isFerried(depositId: Uint8Array): Promise<boolean>;
	getNativeTokenAddress(): Promise<Uint8Array>;
	valutateToken(tokenAddress: Uint8Array, amount: bigint): Promise<bigint>;
}

class L1Api implements L1Interface {
	client!: PublicClient;

	constructor(uri: string) {
		if (uri.startsWith("ws")) {
			this.client = createPublicClient({
				transport: webSocket(uri, { retryCount: 5 }),
			});
		} else if (uri.startsWith("http")) {
			this.client = createPublicClient({
				transport: http(uri, { retryCount: 5 }),
			});
		} else {
			throw new Error("Invalid uri");
		}
	}

	async getDepostiHash(requestId: bigint): Promise<Uint8Array> {
		const value = await this.client.readContract({
			address: MANGATA_CONTRACT_ADDRESS,
			abi: ABI,
			functionName: "deposits",
			blockTag: "finalized",
			args: [requestId],
		});
		const hash = encodeAbiParameters(
			ABI.find((e: any) => e!.name === "deposits")!.outputs!,
			[...(value as any)],
		);
		return hexToU8a(keccak256(hexToU8a(hash)));
	}

	async getLatestRequestId(): Promise<bigint | null> {
		const value = await this.client.readContract({
			address: MANGATA_CONTRACT_ADDRESS,
			abi: ABI,
			functionName: "counter",
			blockTag: "finalized",
		});
		const reqId = BigInt(value as any) - 1n;
		if (reqId < 1n) {
			return null;
		}
		return reqId;
	}

	async getDeposits(rangeStart: bigint, rangeEnd: bigint): Promise<Deposit[]> {
		const selector = "getPendingRequests";
		const contractData = await this.client.readContract({
			address: MANGATA_CONTRACT_ADDRESS,
			abi: ABI,
			functionName: selector,
			args: [rangeStart, rangeEnd],
		});

		return (contractData as any).pendingDeposits.map((deposit: any) => {
			return {
				requestId: deposit.requestId.id,
				depositRecipient: hexToU8a(deposit.depositRecipient),
				tokenAddress: hexToU8a(deposit.tokenAddress),
				amount: deposit.amount,
				timeStamp: deposit.timeStamp,
				ferryTip: deposit.ferryTip,
			};
		});
	}

	async isRolldownDeployed(): Promise<boolean> {
		const code = await this.client.getCode({
			address: MANGATA_CONTRACT_ADDRESS,
			blockTag: "finalized",
		});
		return code !== "0x";
	}
}

class L2Api implements L2Interface {
	api!: ApiPromise;
	keyring!: KeyringPair;

	constructor(api: ApiPromise) {
		this.api = api;
	}

	async valutateToken(
		tokenAddress: Uint8Array,
		amount: bigint,
	): Promise<bigint> {
		const asset: MangataTypesAssetsL1Asset = createL1Asset(
			this.api,
			tokenAddress,
		);
		const tokenId = await this.api.query.assetRegistry.l1AssetToId(asset);
		if (tokenId.isNone) {
			return 0n;
		}

		if (tokenId.unwrap().toNumber() == 0) {
			return amount;
		} else {
			return BigInt(
				(
					await this.api.rpc.xyk.calculate_sell_price(
						tokenId.unwrap(),
						0,
						amount,
					)
				).toString(),
			);
		}
	}

	async getNativeTokenAddress(): Promise<Uint8Array> {
		return (await this.api.query.assetRegistry.idToL1Asset(0))
			.unwrap()
			.asEthereum.toU8a();
	}

	async getBalances(address: Uint8Array): Promise<Map<Uint8Array, bigint>> {
		const assetMapping =
			await this.api.query.assetRegistry.idToL1Asset.entries();
		const chain = getL1ChainType(this.api);

		const mapping: [bigint, Uint8Array][] = assetMapping
			.filter(
				([_key, value]) =>
					(value.unwrap().isEthereum && chain.isEthereum) ||
					(value.unwrap().isArbitrum && chain.isArbitrum),
			)
			.map(([key, value]) => {
				const id = BigInt(key.args[0].toString());
				const address = value.unwrap();
				if (address.isArbitrum && chain.isArbitrum) {
					return [id, value.unwrap().asArbitrum.toU8a()];
				} else if (address.isEthereum && chain.isEthereum) {
					return [id, value.unwrap().asEthereum.toU8a()];
				}
        throw new Error("Invalid chain type");
			});

		const idToL1Asset = new Map<bigint, Uint8Array>(mapping);

		const balances = await this.api.query.tokens.accounts.entries(address);
		const values: [Uint8Array, bigint][] = balances
			.filter(([key, _value]) =>
				idToL1Asset.has(BigInt(key.args[1].toString())),
			)
			.map(([key, value]) => {
				const tokenAddress = idToL1Asset.get(BigInt(key.args[1].toString()))!;
				return [tokenAddress, BigInt(value.free.toString())];
			});

		return new Map<Uint8Array, bigint>(values);
	}

	async getLastProcessedRequestId(): Promise<bigint> {
		const last =
			await this.api.query.rolldown.lastProcessedRequestOnL2(L1_CHAIN);
		return BigInt(last.toString());
	}

	async isExecuted(depositId: bigint): Promise<boolean> {
		return depositId <= (await this.getLastProcessedRequestId());
	}

	async isFerried(depositHash: Uint8Array): Promise<boolean> {
		if (depositHash.length !== 32) {
			throw new Error("depositHash must be exactly 32 bytes long.");
		}
		const chain = getL1ChainType(this.api);
		const status = await this.api.query.rolldown.ferriedDeposits([
			chain,
			depositHash,
		]);
		return status.isSome;
	}
}

function sleep(timeInMilliseconds: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, timeInMilliseconds));
}

function getKeyring(type: KeypairType): Keyring {
	return new Keyring({ type });
}

function getCollator(type: KeypairType, mnemonic: string): KeyringPair {
	const keyring = getKeyring(type);
	return keyring.addFromSeed(hexToU8a(mnemonic));
}

async function getApi(nodeUrl: string): Promise<ApiPromise> {
	const api = await Mangata.instance([nodeUrl]).api();
	await api.isReady;
	return api;
}

async function getNativeL1Update(
	api: ApiPromise,
	encodedData: `0x${string}`,
): Promise<Option<PalletRolldownMessagesL1Update>> {
	return await api.rpc.rolldown.get_native_sequencer_update(
		encodedData.substring(2),
	);
}

async function getEvents(
	apiAt: ApiDecoration<"promise">,
	section: string,
	method: string,
): Promise<FrameSystemEventRecord[]> {
	const events = await apiAt.query.system.events();

	return events.filter(
		(event) => event.event.section === section && event.event.method === method,
	);
}

function isSuccess(events: MangataGenericEvent[]) {
	return events.some(
		(event) =>
			event.section === "system" && event.method === "ExtrinsicSuccess",
	);
}

function print(data: any) {
	console.log(util.inspect(data, { depth: null }));
}

export {
	print,
	sleep,
	getApi,
	getEvents,
	isSuccess,
	getCollator,
	getNativeL1Update,
	L1Api,
	type L1Interface,
	L2Api,
	type L2Interface,
	dummyDeposit,
	Ferry,
	type Deposit,
	getL1ChainType,
};
