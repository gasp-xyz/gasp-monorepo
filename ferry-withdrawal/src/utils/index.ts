import util from "node:util";
import { type ApiPromise, Keyring } from "@polkadot/api";
import { Mangata, type MangataGenericEvent, signTx } from "gasp-sdk";
import { createLogger, transports, format } from "winston";
import { LOG } from "../common/constants.js";

const myFormat = format.printf(({ level, message, label, timestamp }) => {
  return `${timestamp} ${level}: ${message}`;
});

const logger = createLogger({
  format: format.combine(
    format.timestamp(),
    format.colorize(),
    myFormat
  ),
  transports: [new transports.Console()]
});

function sleep(timeInMilliseconds: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, timeInMilliseconds));
}

async function getApi(nodeUrl: string): Promise<ApiPromise> {
	const api = await Mangata.instance([nodeUrl]).api();
	await api.isReady;
	return api;
}

function isSuccess(events: MangataGenericEvent[]) {
	return events.some(
		(event) =>
			event.section === "system" && event.method === "ExtrinsicSuccess",
	);
}

function isEqual(first: Uint8Array, second: Uint8Array): boolean {
  if(first.length !== second.length) {
    return false;
  }
  return first.every((value, index) => value === second[index]);
}

function print(data: any) {
	console.log(util.inspect(data, { depth: null }));
}

export { print, sleep, getApi, isSuccess, isEqual, logger };
