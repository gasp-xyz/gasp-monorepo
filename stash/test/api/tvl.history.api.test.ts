import supertest from 'supertest'
import { describe, expect,it } from 'vitest'

import app from '../../src/app'
import { tokenIDs } from './utils'

describe('/GET tvl-history', () => {
  it.skip('GET /tvl-history - Schema validation', async () => {
    await supertest(app)
      .get(`/tvl-history/0`)
      .expect(200)
      .then((response) => {
        expect(response.body).toMatchSnapshot('Tvl-history')
        expect(response.body.error).toBeUndefined()
        expect(response.body.volumes).toBeDefined()
        expect(response.body.volumes).toBeInstanceOf(Array)
      })
  })

  it.each(tokenIDs)(
    'should return prices for supported pools: %s',
    async (tokenID) => {
      const response = await supertest(app)
        .get('/price-history/' + tokenID)
        .expect(200)
      expect(response.body).to.have.property('prices')
      expect(response.body.error).toBeUndefined()
      expect(response.body.prices).toBeDefined()
      expect(response.body.prices).toBeInstanceOf(Array)
    }
  )

  describe('API errors', () => {
    it('should return error on non existing ID', async () => {
      const errorMessage =
        'this must be one of the following values: 0, 1, 2, 3, 4, 5, 7, 15, 19'

      await supertest(app)
        .get(`/tvl-history/foo`)
        .expect(200)
        .then((response) => {
          expect(response.body).toEqual({ volumes: [] })
        })
    })
  })
})
