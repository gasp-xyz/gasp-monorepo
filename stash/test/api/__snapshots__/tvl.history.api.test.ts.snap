// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/GET tvl-history > GET /tvl-history - Schema validation > Tvl-history 1`] = `
{
  "volumes": [
    [
      1721001600000,
      "935675671.7521595",
    ],
    [
      1721088000000,
      "1341149531.7636049",
    ],
    [
      1721174400000,
      "1494253450.7842066",
    ],
    [
      1721260800000,
      "1275344713.323399",
    ],
    [
      1721347200000,
      "1241480987.3136861",
    ],
    [
      1721433600000,
      "1101145945.3084762",
    ],
    [
      1721520000000,
      "1383297263.302962",
    ],
    [
      1721606400000,
      "1364771293.6166096",
    ],
    [
      1721692800000,
      "1352009603.4200287",
    ],
    [
      1721779200000,
      "1367690311.1749687",
    ],
    [
      1721865600000,
      "1255401206.180484",
    ],
    [
      1721952000000,
      "996243468.4643209",
    ],
    [
      1722038400000,
      "1168690036.995604",
    ],
    [
      1722124800000,
      "1104795669.5428581",
    ],
    [
      1722211200000,
      "1051403659.4419665",
    ],
    [
      1722297600000,
      "1091850338.4185967",
    ],
    [
      1722384000000,
      "732712782.6071895",
    ],
    [
      1722470400000,
      "649417575.6098323",
    ],
    [
      1722556800000,
      "641792152.8920195",
    ],
    [
      1722643200000,
      "599030988.8002836",
    ],
    [
      1722729600000,
      "599354818.4240719",
    ],
    [
      1722816000000,
      "537329683.131297",
    ],
    [
      1722902400000,
      "484122522.91281426",
    ],
    [
      1722988800000,
      "489525693.35823685",
    ],
    [
      1723075200000,
      "467521595.74886113",
    ],
    [
      1723161600000,
      "535261388.27057767",
    ],
    [
      1723248000000,
      "515438146.0829553",
    ],
    [
      1723334400000,
      "520149079.2316735",
    ],
    [
      1723420800000,
      "510262801.0495691",
    ],
    [
      1723507200000,
      "541808904.0358355",
    ],
    [
      1723593600000,
      "521873181.732515",
    ],
    [
      1723680000000,
      "549020909.6532067",
    ],
    [
      1723766400000,
      "492134772.08386374",
    ],
    [
      1723852800000,
      "489763435.47502166",
    ],
    [
      1723939200000,
      "495423862.323143",
    ],
    [
      1724025600000,
      "494551333.1181789",
    ],
    [
      1724112000000,
      "493956084.266465",
    ],
    [
      1724198400000,
      "475125379.54989547",
    ],
    [
      1724284800000,
      "490301001.7519652",
    ],
    [
      1724371200000,
      "474892802.4996398",
    ],
    [
      1724457600000,
      "494095514.79178995",
    ],
    [
      1724544000000,
      "498323066.9177997",
    ],
    [
      1724630400000,
      "498563536.4551976",
    ],
    [
      1724716800000,
      "484565292.1247067",
    ],
    [
      1724803200000,
      "431450970.63995504",
    ],
    [
      1724889600000,
      "436520937.8588748",
    ],
    [
      1724976000000,
      "434283085.1444852",
    ],
    [
      1725062400000,
      "437297996.66762364",
    ],
    [
      1725148800000,
      "431241831.0319696",
    ],
    [
      1725235200000,
      "411590007.46850747",
    ],
    [
      1725321600000,
      "420608034.06748927",
    ],
    [
      1725408000000,
      "399884708.96008575",
    ],
    [
      1725494400000,
      "399118052.29553497",
    ],
    [
      1725580800000,
      "382646807.5090376",
    ],
    [
      1725667200000,
      "361872476.93431544",
    ],
    [
      1725753600000,
      "367947547.11974967",
    ],
    [
      1725840000000,
      "363862663.7759709",
    ],
    [
      1725926400000,
      "371773372.3902221",
    ],
    [
      1726012800000,
      "373532130.0813272",
    ],
    [
      1726099200000,
      "366817554.2059798",
    ],
    [
      1726185600000,
      "367880871.4275921",
    ],
    [
      1726272000000,
      "380185050.0983515",
    ],
    [
      1726358400000,
      "372940668.30490005",
    ],
    [
      1726444800000,
      "352031453.45089674",
    ],
    [
      1726531200000,
      "342895295.2829355",
    ],
    [
      1726617600000,
      "359281498.8706001",
    ],
    [
      1726704000000,
      "496215237.3714284",
    ],
    [
      1726790400000,
      "362114997.9242417",
    ],
    [
      1726876800000,
      "375682923.6416214",
    ],
    [
      1726963200000,
      "384052248.19958305",
    ],
    [
      1727049600000,
      "373191637.470824",
    ],
    [
      1727136000000,
      "372975404.2096534",
    ],
    [
      1727222400000,
      "370732904.20607835",
    ],
    [
      1727308800000,
      "361417458.27883637",
    ],
    [
      1727395200000,
      "364993407.1455418",
    ],
    [
      1727481600000,
      "373163306.15585625",
    ],
    [
      1727568000000,
      "370677463.897763",
    ],
    [
      1727654400000,
      "367719813.47896266",
    ],
    [
      1727740800000,
      "358958788.1432284",
    ],
    [
      1727827200000,
      "338730535.4872396",
    ],
    [
      1727913600000,
      "466601873.9805762",
    ],
    [
      1728000000000,
      "559433968.8209311",
    ],
  ],
}
`;
